### PR TITLE
Paged llama model

### PR DESCRIPTION
**Pull request by JRosenkranz**
This PR is part of a stack of PRs related to paged attention + speculative decoding:

- [x]  Paged Attention KVCache (#8)
- [ ]  Paged Model
- [ ] Speculative generation

Full implementation of the above can be found here: https://github.com/foundation-model-stack/fms-extras/pull/7

In this PR, a paged LLaMA model implementation is introduced which utilizes the paged attention kv-cache manager and paged attention kernels in MHA (from #8). This PR also includes a simple generation utility and a script to demonstrate generation.

### CODE REVIEW

File: fms_extras/models/paged_llama.py [hunk 1]
```diff
@@ -0,0 +1,844 @@
...
+    pad_id: int = -1
+    hidden_grow_factor: float = 8 / 3
+    multiple_of: int = 256
+    activation_fn: str = "swish"
+    p_dropout: float = 0.0
+    max_expected_seq_len: int = 4096
+    ntk_scaling: bool = False
+
+
+class PagedMultiHeadAttention(nn.Module):
```
Comment (nairbv) [inline fms_extras/models/paged_llama.py:59]: this is mostly copy/paste from fms llama right? I'm not reviewing in detail since I'm assuming it's basically all the same, just watching out for where I expect differences

File: fms_extras/models/paged_llama.py [hunk 2]
```diff
@@ -0,0 +1,844 @@
...
+models.register_model(
+    _architecture_name, "micro", _llama_factory_factory(_micro_char_config)
+)
+models.register_model(_architecture_name, "7b", _llama_factory_factory(_7b_config))
+models.register_model(_architecture_name, "13b", _llama_factory_factory(_13b_config))
+models.register_model(_architecture_name, "70b", _llama_factory_factory(_70b_config))
+
+
+def _rename_weights_to_fms(orig_sd):
+    replacements = [
```
Comment (nairbv) [inline fms_extras/models/paged_llama.py:711]: i won't block this PR on it, and I know there are some small differences, but I suspect there's something better we could be doing with these adapters to avoid so much repetition.

File: fms_extras/models/paged_llama.py [hunk 3]
```diff
@@ -0,0 +1,844 @@
...
+def _llama_factory_factory(config):
+    def factory(**kwargs):
+        return PagedLLaMA(config, **kwargs)
+
+    return factory
+
+
+models.register_model(
+    _architecture_name, "micro", _llama_factory_factory(_micro_char_config)
+)
```
Comment (nairbv) [inline fms_extras/models/paged_llama.py:704]: do we have any tests that would enforce consistency checks between this and the original llama? they should behave the same right? that might be an easy test to add even without the whole signature framework, since we could just compare output of micro llama with micro llama_paged and random/arbitrary input.

File: fms_extras/utils/cache/__init__.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
+import abc
+import dataclasses
+from typing import List, Optional, Tuple
+
+import torch
+
+
+@dataclasses.dataclass
+class CacheDataLayer(metaclass=abc.ABCMeta):
```
Comment (nairbv) [inline fms_extras/utils/cache/__init__.py]: i think we still have a single-use interface here?

File: fms_extras/models/paged_llama.py (hunk 3 above)
Comment (JRosenkranz) [reply fms_extras/models/paged_llama.py:704]: yes, i can add this

File: fms_extras/utils/cache/__init__.py (hunk 4 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/__init__.py]: removed in the cache PR, will fix this one.

File: fms_extras/utils/generation.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        top_k: int
+            only search among top k tokens
+        do_sample: bool
+            multinomial sampling. False for greedy.
+        decode_model: Callable or nn.Module, optional
+            a model to used specifically for decode step. If not given, the model input will be used for decode step
+
+    Returns:
+    Tuple[torch.Tensor, int, int]
+        the resulting output tokens, the number of new tokens generated, and the time it took per token in seconds
```
Comment (nairbv) [inline fms_extras/utils/generation.py]: the third term returned is `(end_time - start_time)` so I think this should be float seconds, not int seconds per token

File: fms_extras/utils/generation.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        input_ids = next_input[:, -max_seq_len:]
+
+        # compute the mask
+        if i == 0:
+            is_pad = input_ids == 0
+            mask = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+            kwargs["mask"] = mask.tril(diagonal=0)
+        else:
+            if i == 1:
+                start_time = time.time()
```
Comment (nairbv) [inline fms_extras/utils/generation.py]: shouldn't we include the first pass in the timer?

File: fms_extras/utils/generation.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        kwargs["cache_data"] = cache_data
+        kwargs["position_ids"] = cache_data.compute_position_ids(
+            num_tokens_per_sequence
+        )
+
+        if i == 0:
+            logits, _ = model(input_ids, **kwargs)
+        else:
+            # cudagraph requires static shapes
+            if cudagraphs:
```
Comment (nairbv) [inline fms_extras/utils/generation.py:122]: can any of the larger nested chunks of code be extracted out into helper functions? Overall this function is a bit long with a lot of nesting so I think we can find some parts to refactor out to help with readability.

File: scripts/paged_speculative_inference.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+import time
+
+import torch
+import torch._inductor.config
+from fms.models import get_model
+from fms.utils import generation, tokenizers
+from torch import distributed as dist
+
+import fms_extras.models.paged_llama
+from fms_extras.utils.generation import paged_generate
```
Comment (nairbv) [inline scripts/paged_speculative_inference.py:13]: other than using paged_generate, is everything else in this file copy/paste? it would be nice to improve code reuse for these kinds of scripts, maybe by modifying the ones in fms.

File: fms_extras/utils/generation.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+import time
+from typing import Any, Callable, List, MutableMapping, Optional, Union
+
+import torch
+import torch.nn.functional as F
+
+from fms_extras.utils.cache.paged import PagedKVCacheManager
+
+
+def paged_generate(
```
Comment (nairbv) [inline fms_extras/utils/generation.py:38]: can we add a unit test for this function?

Comment (JRosenkranz) [reply fms_extras/utils/generation.py:38]: yes, would require GPU, but should be fine to add

File: scripts/paged_speculative_inference.py (hunk 8 above)
Comment (JRosenkranz) [reply scripts/paged_speculative_inference.py:13]: This was adapted from the one in `fms`, but includes a warmup, some extra metrics, initializes the kv-cache, as well as calls different functions for generate (including speculator in later PR). Might make sense at some point to add some generic script for inference in `fms` that we could extend with our own custom functions to create other inference scripts. For now, I think this should be kept separate as there is enough differences.

File: fms_extras/utils/generation.py (hunk 6 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: we could, though we were only timing the decode step. Let me check and get back to you how we would like to time.

File: fms_extras/utils/generation.py (hunk 9 above)
Comment (nairbv) [reply fms_extras/utils/generation.py:38]: why does it require GPU? in fms we have generate tests like this:

https://github.com/foundation-model-stack/foundation-model-stack/blob/main/tests/utils/test_generate.py#L27

Comment (JRosenkranz) [reply fms_extras/utils/generation.py:38]: Currently, paged kernels only work on cuda

File: fms_extras/utils/generation.py (hunk 5 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: fixed in latest commit

File: fms_extras/utils/generation.py (hunk 7 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py:122]: added a function for this to pad to right for cudagraphs in latest commit

File: fms_extras/utils/generation.py (hunk 9 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py:38]: added test which tests the output of a micro model in paged_generate vs a micro model with fms generate in latest commit

File: fms_extras/models/paged_llama.py (hunk 3 above)
Comment (JRosenkranz) [reply fms_extras/models/paged_llama.py:704]: added in latest commit

File: fms_extras/utils/generation.py (hunk 5 above)
Comment (nairbv) [reply fms_extras/utils/generation.py]: I think it still should be total time not per-token time though right?

File: fms_extras/utils/generation.py (hunk 6 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: Added timing for the first token and the subsequent decode tokens in latest commit

File: fms_extras/utils/generation.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        else:
+            kwargs["mask"] = None
+
+        # get the cache data and position ids if using cache
+        if sequence_ids is None:
+            num_tokens_per_sequence = torch.count_nonzero(input_ids.T, dim=0).tolist()
+        else:
+            num_tokens_per_sequence = [1 for _ in range(input_ids.size(0))]
+
+        cache_data = kv_cache_manager.allocate_tokens(
```
Comment (nairbv) [inline fms_extras/utils/generation.py]: i thouguht we only want to allocate a cache_data object with the first token (if not already in kwargs)? isn't this supposed to be like a kv cache?

File: fms_extras/utils/generation.py (hunk 7 above)
Comment (nairbv) [reply fms_extras/utils/generation.py:122]: is there any more that can be refactored out? It seems like we could even have an entire update_kwargs function such that the generate code was really just handling the generation loop, not managing position IDs and such

File: fms_extras/utils/generation.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            num_tokens_per_sequence, sequence_ids
+        )
+
+        sequence_ids = cache_data.sequence_ids
+
+        kwargs["cache_data"] = cache_data
+        kwargs["position_ids"] = cache_data.compute_position_ids(
+            num_tokens_per_sequence
```
Comment (nairbv) [inline fms_extras/utils/generation.py]: I think this was a previous PR that maybe I missed when revisiting but I'm still finding this confusing. The doc for this parameter says "number of tokens to be added to each sequence" ... that implies it would be `max_new_tokens` but obviously isn't. I think we still need to clean up the comments and/or parameter names. I guess this is non-pad tokens per sequence?

File: fms_extras/utils/generation.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        if i == 0:
+            is_pad = input_ids == 0
+            mask = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+            kwargs["mask"] = mask.tril(diagonal=0)
+        else:
+            kwargs["mask"] = None
+
+        # get the cache data and position ids if using cache
+        if sequence_ids is None:
+            num_tokens_per_sequence = torch.count_nonzero(input_ids.T, dim=0).tolist()
```
Comment (nairbv) [inline fms_extras/utils/generation.py]: i still don't love using non-zero to find non-pad tokens. A prompt could contain a 0, which means 'unk' in the llama tokenizer, and could mean something else with some other tokenizer.

File: fms_extras/utils/generation.py (hunk 10 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: So allocation here is logical allocation, the kv-cache tensor was fully allocated when it was initialized in the kv_cache_manager. When you allocate tokens, you are essentially reserving blocks from the already pre-allocated space.

File: fms_extras/utils/generation.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        decode_model: Callable or nn.Module, optional
+            a model to used specifically for decode step. If not given, the model input will be used for decode step
+        cudagraphs: bool
+            if True, model is using cudagraphs and input will be padded, otherwise no cudagraphs is used and input is
+            not padded
+
+    Returns:
+    Tuple[torch.Tensor, int, float, float]
+        the resulting output tokens, the number of new tokens generated, the time to first token in seconds and the
+        total decode time
```
Comment (nairbv) [inline fms_extras/utils/generation.py]: I'm not sure "decode time" clearly indicates that we mean total time to generate subsequent tokens after kv-cache is populated

File: fms_extras/utils/generation.py (hunk 11 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: This is unrelated to max_new_tokens, it is just creating a new position_ids object based on the current sequence length (not filled) in the cache and the new token to add per decode step.

As an example:

current_sequence_lengths = [50, 99, 16]
num_tokens_per_sequence = [5, 5, 5]

This will give you position_ids

[
     [50, 51, 52, 53, 54], 
     [99, 100, 101, 102, 103],
     [16, 17, 18, 19, 20]
]

If this is prefill, the current sequence length is essentially None, and a padded position_ids is created based on the prompt length

I can add more documentation to this function if it will help? Also, technically we could just make the assumption when you allocate_tokens that the num_tokens_per_sequence should be used for this computation, in which case we can remove the position_ids computation and just have it as a part of the dataclass

Comment (nairbv) [reply fms_extras/utils/generation.py]: right, I know it's not really related to 'max new tokens', I'm just saying the doc implies that t is. I think we need to make the names and comments make sense.

File: fms_extras/utils/generation.py (hunk 12 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: Because we take in the sequence (not padded), I can extract the num_tokens_per_sequence pretty easily just using the list of input_ids directly. For the mask, I can create it based on this `num_tokens_per_sequence`. I think this should simplify things, and this way we make no assumptions about 0s in the inputs.

File: fms_extras/utils/generation.py (hunk 13 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: Fixed in latest commit

File: fms_extras/utils/generation.py (hunk 12 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: Fixed in latest commit, no assumptions made about token=0

File: fms_extras/utils/generation.py (hunk 7 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py:122]: cleaned the generate up a bit and removed the compute_position_ids. The preprocessing for model_kwargs now should be pretty minimal

File: fms_extras/utils/generation.py (hunk 11 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: updated and renamed to `model_input_lengths`

File: fms_extras/utils/generation.py (hunk 5 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: fixed in latest commit: `the cumulative time of all decode steps in seconds`

File: fms_extras/utils/generation.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+        decode_model: Callable or nn.Module, optional
+            a model to used specifically for decode step. If not given, the model input will be used for decode step
+        cudagraphs: bool
+            if True, model is using cudagraphs and input will be padded, otherwise no cudagraphs is used and input is
+            not padded
+
+    Returns:
+    Tuple[torch.Tensor, int, float, float]
+        the resulting output tokens, the number of new tokens generated, the time to first token in seconds and the
+        cumulative time of all decode steps in seconds
```
Comment (nairbv) [inline fms_extras/utils/generation.py]: this still isn't correct, right?

Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: I am getting the total time of the generation loop post prefill, so seems ok?

Comment (nairbv) [reply fms_extras/utils/generation.py]: your code has:
```
        if i == 1:
            start_time = time.time()
```
so it's only the subsequent tokens as-written?

Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: yes, we return the time to first token (which is the first pass), then the cumulative of all other forward passes coming after the first pass

Comment (nairbv) [reply fms_extras/utils/generation.py]: "the cumulative time of *all* decode steps in seconds" doesn't mean the same thing as "the cumulative of all *other* forward passes coming after the first pass"

File: fms_extras/utils/generation.py [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+            next_val = torch.multinomial(probs, num_samples=1)
+        else:
+            next_val = torch.argmax(logits, dim=-1).unsqueeze(0).t()
+
+        result = torch.cat((result, next_val), dim=-1)
+
+        next_input = next_val
+
+        if i == 0:
+            ttft = time.time() - start_time
```
Comment (nairbv) [inline fms_extras/utils/generation.py:147]: also it might be easier to follow if we move the `start_time = time.time()` to here, right after setting ttft, instead of under `if i == 1`. Maybe even just move everything under `i==1` to here?

File: fms_extras/utils/generation.py (hunk 14 above)
Comment (JRosenkranz) [reply fms_extras/utils/generation.py]: Ah yes, it's including the cache management and setup in the loop

**Review by nairbv — APPROVED** [review]
